src/main/java/com/gooddata/http/client/GoodDataHttpClient.java

```diff
@@ -147,29 +183,23 @@ private HttpResponse handleResponse(final HttpHost httpHost, final HttpRequest r
         }
         EntityUtils.consume(originalResponse.getEntity());
 
-        final Lock authLock = (Lock) context.getAttribute(LOCK_AUTH);
-        final boolean entered = authLock != null ? authLock.tryLock() : true;
+        final boolean entered = authLock.tryLock();
 
         if (entered) {
             try {
                 //only one thread requiring authentication will get here.
-                final ReadWriteLock rwLock = (ReadWriteLock) context.getAttribute(LOCK_RW);
-                Lock writeLock = null;
-                if (rwLock != null) {
-                    writeLock = rwLock.writeLock();
-                    writeLock.lock();
-                }
+                final Lock writeLock = rwLock.writeLock();
+                writeLock.lock();
                 boolean doSST = true;
                 try {
-                    if (challenge == GoodDataChallengeType.TT) {
-                        if (this.refreshTt(httpHost)) {
+                    if (challenge == GoodDataChallengeType.TT && sst != null) {
+                        if (refreshTt()) {
                             doSST = false;
                         }
                     }
                     if (doSST) {
-                        final String sst = sstStrategy.obtainSst();
-                        CookieUtils.replaceSst(sst, context, httpHost.getHostName());
-                        if (!refreshTt(httpHost)) {
+                        sst = sstStrategy.obtainSst(httpClient, authHost);
```

Comment:
TT is send with every request to GoodData resource. When the TT expires, the server sends TT challenge and the client has to obtain a new TT. Yes, it may happen SST expires as well (but it has longer life -- like 24 hours or so).